File: apps/customer-mobile/shared/src/commonTest/kotlin/com/effyshopping/customer/mobile/features/payment/PaymentWireContractTest.kt
package com.effyshopping.customer.mobile.features.payment

import com.effyshopping.customer.mobile.commerce.contract.BillingDetailsDTO
import com.effyshopping.customer.mobile.commerce.contract.ListPaymentMethodsResponse
import com.effyshopping.customer.mobile.commerce.contract.PaymentMethodDTO
import kotlinx.serialization.json.Json
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * The Kotlin half of the 051 payment wire contract.
 *
 * ⚠ The literals below are duplicated BYTE-FOR-BYTE from the Go half at
 * apis/core-api/internal/features/checkout/payment_wire_contract_test.go. Neither side generates or
 * imports the fixture — a shared literal moves WITH a bug, while two hand-kept copies make a
 * divergence show up as a failure. That is the 028 pattern, and it was proved there by breaking it
 * two ways.
 *
 * ⚠ WHY THIS EXISTS. 027 R13 lost days because Kotlin serialised a count as `Double`, so the wire
 * carried `1.0` and Go's encoding/json refused it into an int. Every unit test on both sides passed
 * throughout, because the fakes spoke Kotlin at both ends and never crossed the wire. A card expiry is
 * a count; this test is the thing standing between that lesson and its fourth recurrence.
 */
class PaymentWireContractTest {

    private val json = Json { ignoreUnknownKeys = true; explicitNulls = false }

    companion object {
        // Byte-identical to paymentMethodWire in payment_wire_contract_test.go.
        const val PAYMENT_METHOD =
            """{"id":"pm_123","brand":"visa","last4":"4242","expMonth":4,"expYear":2028,"isDefault":true,"usable":true}"""

        // Byte-identical to paymentMethodUnusableWire.
        const val PAYMENT_METHOD_UNUSABLE =
            """{"id":"pm_456","brand":"mastercard","last4":"8210","expMonth":7,"expYear":2026,"isDefault":false,"usable":false,"unusableReason":"This card has expired."}"""

        // Byte-identical to billingDetailsWire.
        const val BILLING_DETAILS =
            """{"name":"Jane Smith","email":"jane@example.com","address":{"line1":"1 Test St","line2":"","city":"Richmond","state":"VIC","postalCode":"3121","country":"AU"}}"""
    }

    @Test
    fun `parses a usable payment method from the servers exact bytes`() {
        val pm = json.decodeFromString<PaymentMethodDTO>(PAYMENT_METHOD)
        assertEquals("pm_123", pm.id)
        assertEquals("visa", pm.brand)
        assertEquals("4242", pm.last4)
        assertEquals(4L, pm.expMonth)
        assertEquals(2028L, pm.expYear)
        assertTrue(pm.isDefault)
        assertTrue(pm.usable)
        // A usable card omits the reason entirely; the field is absent, not empty.
        assertNull(pm.unusableReason)
    }

    @Test
    fun `an unusable card always carries a reason a shopper can read`() {
        val pm = json.decodeFromString<PaymentMethodDTO>(PAYMENT_METHOD_UNUSABLE)
        assertFalse(pm.usable)
        // ⚠ FR-023. A card shown as unusable with no reason is a refusal the shopper cannot act on,
        // which is indistinguishable from a bug.
        assertEquals("This card has expired.", pm.unusableReason)
    }

    /**
     * ⚠ THE 027 R13 GUARD.
     *
     * `expMonth` and `expYear` must round-trip as integers. Re-encoding and comparing against the
     * server's exact bytes is what catches a Double: `4.0` would serialise back as `4.0` and fail here,
     * where an assertion on the decoded VALUE alone would happily pass, because 4.0 == 4.
     */
    @Test
    fun `expiry round-trips as an integer and never as a float`() {
        val pm = json.decodeFromString<PaymentMethodDTO>(PAYMENT_METHOD)
        val reEncoded = json.encodeToString(PaymentMethodDTO.serializer(), pm)
        assertTrue(
            reEncoded.contains("\"expMonth\":4") && reEncoded.contains("\"expYear\":2028"),
            "expiry serialised with a decimal point — this is the 027 R13 defect: $reEncoded",
        )
        assertFalse(
            reEncoded.contains("4.0") || reEncoded.contains("2028.0"),
            "a count crossed the wire as a float: $reEncoded",
        )
    }

    @Test
    fun `parses the billing details the client must send back at confirmation`() {
        val bd = json.decodeFromString<BillingDetailsDTO>(BILLING_DETAILS)
        assertEquals("Jane Smith", bd.name)
        assertEquals("1 Test St", bd.address.line1)
        assertEquals("3121", bd.address.postalCode)
        // ⚠ The whole point of the feature: the shopper is never asked for a country, so the server
        // supplies one and the client must receive it. An absent country is what let the provider guess.
        assertEquals("AU", bd.address.country)
    }

    @Test
    fun `an empty list parses as no kept cards`() {
        val res = json.decodeFromString<ListPaymentMethodsResponse>("""{"paymentMethods":[]}""")
        assertTrue(res.paymentMethods.isEmpty())
    }
}

File: apis/core-api/internal/features/checkout/payment_wire_contract_test.go
package checkout

import (
	"testing"
)

// ⚠ THE WIRE CONTRACT for the 051 payment shapes. This pins the exact JSON the hot path emits so a
// field rename or a type drift fails HERE, not in a shopper's payment. The customer-mobile
// PaymentWireContractTest.kt parses the SAME literals, BYTE-IDENTICAL — the two halves are kept in sync
// by hand, which is the 028 pattern and the reason a silent `json:"..."` rename cannot survive.
//
// ⚠ THE INVARIANT THIS EXISTS FOR: `expMonth` and `expYear` are JSON NUMBERS WITHOUT A DECIMAL POINT.
// 027 R13 lost days to Kotlin serialising an integer as `1.0` where Go's encoding/json refuses it into
// an int — three stacked defects, found only by querying the database directly. The TS side declares
// them `WireInt` (@asType integer) so the generated Kotlin is a Long rather than a Double; this is the
// Go end of the same guarantee.

const paymentMethodWire = `{"id":"pm_123","brand":"visa","last4":"4242","expMonth":4,"expYear":2028,"isDefault":true,"usable":true}`

const paymentMethodUnusableWire = `{"id":"pm_456","brand":"mastercard","last4":"8210","expMonth":7,"expYear":2026,"isDefault":false,"usable":false,"unusableReason":"This card has expired."}`

const billingDetailsWire = `{"name":"Jane Smith","email":"jane@example.com","address":{"line1":"1 Test St","line2":"","city":"Richmond","state":"VIC","postalCode":"3121","country":"AU"}}`

func TestWireContract_PaymentMethod(t *testing.T) {
	got := mustJSON(t, paymentMethodBody{
		ID: "pm_123", Brand: "visa", Last4: "4242",
		ExpMonth: 4, ExpYear: 2028, IsDefault: true, Usable: true,
	})
	if got != paymentMethodWire {
		t.Errorf("payment method wire drift:\n got  %s\n want %s", got, paymentMethodWire)
	}
}

// ⚠ A usable card omits `unusableReason` entirely (omitempty); an unusable one MUST carry it. A card
// shown as unusable with no reason is a refusal the shopper cannot act on, which FR-023 forbids.
func TestWireContract_PaymentMethodUnusableCarriesItsReason(t *testing.T) {
	got := mustJSON(t, paymentMethodBody{
		ID: "pm_456", Brand: "mastercard", Last4: "8210",
		ExpMonth: 7, ExpYear: 2026, IsDefault: false,
		Usable: false, UnusableReason: "This card has expired.",
	})
	if got != paymentMethodUnusableWire {
		t.Errorf("unusable payment method wire drift:\n got  %s\n want %s", got, paymentMethodUnusableWire)
	}
}

func TestWireContract_BillingDetails(t *testing.T) {
	got := mustJSON(t, billingDetailsBody{
		Name:  "Jane Smith",
		Email: "jane@example.com",
		Address: billingAddressBody{
			Line1: "1 Test St", City: "Richmond", State: "VIC",
			PostalCode: "3121", Country: "AU",
		},
	})
	if got != billingDetailsWire {
		t.Errorf("billing details wire drift:\n got  %s\n want %s", got, billingDetailsWire)
	}
}

// ⚠ THE 027 R13 GUARD, asserted directly rather than left implicit in the shape above.
//
// A card expiry is a count, not a measurement. If either field ever serialises with a decimal point,
// the Kotlin client and the Go server have disagreed about what an integer is — and that disagreement
// is invisible until a real device hits a real endpoint.
func TestWireContract_ExpiryIsAnIntegerNotAFloat(t *testing.T) {
	got := mustJSON(t, paymentMethodBody{
		ID: "pm_1", Brand: "visa", Last4: "0000", ExpMonth: 1, ExpYear: 2030, Usable: true,
	})
	want := `{"id":"pm_1","brand":"visa","last4":"0000","expMonth":1,"expYear":2030,"isDefault":false,"usable":true}`
	if got != want {
		t.Errorf("expiry wire drift (a decimal point here is the 027 R13 defect):\n got  %s\n want %s", got, want)
	}
}

// ⚠ THE FIELD THAT MUST NEVER APPEAR. The provider customer reference identifies a provider record and
// no surface has any use for it (data-model § 1). Asserting the WHOLE serialised object rather than
// checking for a substring is what makes this catch a future field added by someone who did not read
// the rule.
func TestWireContract_IntentResponseCarriesNoProviderCustomerID(t *testing.T) {
	got := mustJSON(t, createIntentResponse{
		OrderID: "o1", OrderNumber: "EFY-1", ClientSecret: "cs_1",
		PublishableKey: "pk_test", GrandTotalAmount: "14.60", Currency: "AUD",
	})
	want := `{"orderId":"o1","orderNumber":"EFY-1","clientSecret":"cs_1","publishableKey":"pk_test","grandTotalAmount":"14.60","currency":"AUD"}`
	if got != want {
		t.Errorf("intent response wire drift:\n got  %s\n want %s\n(a new field here needs a reason; stripe_customer_id never does)", got, want)
	}
}
